src/OpenTelemetry/Batch.cs

```diff
@@ -90,7 +91,11 @@ public void Dispose()
                 // Drain anything left in the batch.
                 while (this.circularBuffer.RemovedCount < this.targetCount)
                 {
-                    this.circularBuffer.Read();
+                    T item = this.circularBuffer.Read();
+                    if (typeof(T) == typeof(LogRecord))
```

Comment:
Maybe use a `boolean` field like `isLogBatch` to decide if we need to return the item to the pool instead of checking it for each item of the batch?

Reply:
That would actual perform worse! What is going on here... JIT magic 🧙‍♂️ What happens is when `Batch<LogRecord>` is JITted, all of these type checks become constant things. For `T`s != `LogRecord` they are completely removed. When `T` == `LogRecord` the checks are removed and the logic is just inlined.

Here is a [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4a9BsXKDREqXwDCEfB3EAbGFCYGAbuLAxcs6nIDMC0g3UMA3jQbuFd4BAi6GAWQAKAEoXNw8I43YGcQwYfAYAXgYRGAB3Bi4IAHMBGEgoABMQqwiIgCFsDDAWAB4s3PzoQoA+BmAqmqSU9MDY+ODSsvcO6pY+ABFxXA4IXBgS8OHiAHYYuPwNfWwRAFUOdWxdfUKh9wBfGkvrWjtcDCghMAxMnLyCwppXagjiLx8/OptnsDkcTi4GNkYBhpAwOFBxFE4gx5jCGNcIksGABtfzQlgQQoASW0ukCeIwBOJpIA8hwJBARLg+AA5CBEkS6cQibnZYIAXSxfwUKEcwKEHBCWO+w3clOmWxgO32h2OMEK3QeQhkWOu11sKK1zwYAEFnojYgBPL5XGg0bEAKViAHEYKkEWBAhhLRwYBAAGaBBrvZrBAX2p0YV3usxen1+wNmiSmb1hwW3Q2PY2VMa1AAqbRADCJUxmc14+htP3c8MRVRgDFg2EKjN0loYefW8TOS2FOZqgU7/XwwSWMrK8uZw+6w7OGPcvbsxFFpdm8yl1bCm484n9DDjvoDg9CiWS3sPgeDTSKYax7nHssCQbe18KwUCEGAACt8hhgpO+GHYJFR2CVFm3C4lj1IA) attempting to demo the concept.

In the final JIT Asm notice...

For `Batch<LogRecord>` the logic is being executed:

```
Batch`1[[LogRecord, _]].Dispose()
    L0000: sub rsp, 0x28
    L0004: xor eax, eax
    L0006: mov [rsp+0x20], rax
    L000b: movzx ecx, byte ptr [rcx]
    L000e: mov [rsp+0x20], cl
    L0012: lea rcx, [rsp+0x20]
    L0017: call 0x00007ffd192404c0
    L001c: nop
    L001d: add rsp, 0x28
    L0021: ret
```

For `Batch<Activity>` the logic is completely removed:

```
Batch`1[[Activity, _]].Dispose()
    L0000: ret
```

Reply:
Ahh that's great!